flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerTest.java

```diff
@@ -389,49 +388,24 @@ public void testGoToFinishedShutsDownCheckpointingComponents() throws Exception
     }
 
     @Test
-    public void testTransitionToStateCallsOnEnter() throws Exception {
+    public void testTransitionToStateCallsOnLeave() throws Exception {
         final DeclarativeScheduler scheduler =
                 new DeclarativeSchedulerBuilder(createJobGraph(), mainThreadExecutor).build();
 
-        final LifecycleMethodCapturingState firstState = new LifecycleMethodCapturingState();
+        final LifecycleMethodCapturingState.Factory firstStateFactory =
+                new LifecycleMethodCapturingState.Factory();
+        final DummyState.Factory secondStateFactory = new DummyState.Factory();
 
-        scheduler.transitionToState(firstState);
-        assertThat(firstState.onEnterCalled, is(true));
-        assertThat(firstState.onLeaveCalled, is(false));
-        firstState.reset();
-    }
-
-    @Test
-    public void testTransitionToStateCallsOnLeave() throws Exception {
-        final DeclarativeScheduler scheduler =
-                new DeclarativeSchedulerBuilder(createJobGraph(), mainThreadExecutor).build();
+        scheduler.transitionToState(firstStateFactory);
 
-        final LifecycleMethodCapturingState firstState = new LifecycleMethodCapturingState();
-        final DummyState secondState = new DummyState();
+        final LifecycleMethodCapturingState firstState =
+                (LifecycleMethodCapturingState) scheduler.getState();
```

Comment:
you could simplify this bit by having a test factory that accepts an constructed state. That's usually a nicer way to deal with factories.